gtsfm/view_graph_estimator/view_graph_estimator_base.py

```diff
@@ -0,0 +1,123 @@
+"""Implements a base class for ViewGraph estimation.
+
+Estimating the ViewGraph can be done trivially by adding all the two-view estimates into a ViewGraph data structure.
+The purpose of this class, however, is to define an API for more sophisticated methods for estimating a ViewGraph 
+that include filtering or optimizing the two-view estimates.
+
+Authors: Akshay Krishnan, Ayush Baid
+"""
+import abc
+from typing import Dict, Optional, List, Tuple
+
+import dask
+import numpy as np
+from dask.delayed import Delayed
+from gtsam import Cal3Bundler, PinholeCameraCal3Bundler, Rot3, Unit3
+
+import gtsfm.utils.geometry_comparisons as comp_utils
+from gtsfm.evaluation.metrics import GtsfmMetric, GtsfmMetricsGroup
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.common.view_graph import ViewGraph
+
+METRIC_GROUP = "view_graph"
+
+
+class ViewGraphEstimatorBase(metaclass=abc.ABCMeta):
```

Comment:
What's the motivation behind making this an abstract class where filtering techniques are implemented as derived classes as opposed to a class with the filtering techniques as methods?

Reply:
what is the motivation for a class with filtering techniques as methods? we would not use more than one of them at any time. 
The abstract class is the same design we follow elsewhere in gtsfm. For example, for detection we don't have one class with different methods for SIFT and deep front end. To me it captures the idea of the derived classes implementing different ways to achieve a common goal, using the same API.

Reply:
I assuming that we'd only ever want to keep track of a single ViewGraph (or multiple ViewGraphs that have been segmented from some global ViewGraph), as there is a lot of memory overhead as the number of views increase. 

Therefore, I think this would make more sense if the `ViewGraphEstimator.run`  took as input a ViewGraph and then output a filtered ViewGraph. This would make it more obvious how the outputs of the different estimators fit together.

Reply:
Where would you create the ViewGraph that is passed to the ViewGraphEstimator ? It seems odd to have a class for estimating a ViewGraph receive a ViewGraph as input. That API would surely surprise a reader?

Reply:
According to your class description: "Estimating the ViewGraph can be done trivially by adding all the two-view estimates into a ViewGraph data structure. The purpose of this class, however, is to define an API for more sophisticated methods for estimating a ViewGraph that include filtering or optimizing the two-view estimates."

Creating the ViewGraph is trivial. The real purpose of this class is to filter the ViewGraph. You could change the name if you'd like.

Reply:
Two reasons why I dont think we should pass in a ViewGraph: 
- We expect a ViewGraphEstimator to compute what is semantically defined as a  “ViewGraph”. These are good edges (say, cycle consistent) for averaging .
- The ViewGraph struct should not be used to store something we don’t semantically call a “ViewGraph”. This will lead to confusion. 
- More overhead (even if its two lines) in MVO

Only reason why we should pass in a ViewGraph: 
- Cleaner API for the ViewGraph class

Reply:
Could you suggest better names if you think we should pass in a ViewGraph? Maybe ViewGraphRefiner.

Reply:
> We expect a ViewGraphEstimator to compute what is semantically defined as a “ViewGraph”. These are good edges (say, cycle consistent) for averaging.

It would still be computing a ViewGraph, just from another ViewGraph. You could change it to something like `ViewGraphOperator` or something like that idk. 

> The ViewGraph struct should not be used to store something we don’t semantically call a “ViewGraph”. This will lead to confusion.

I'm not following.

Reply:
Maybe this is also a change to be made when we use the ViewGraph beyond filtering two-view edges.

Reply:
Re the abstract class design: this is what we follow everywhere in `GtSFM`. Coding to interface is the way to go, even if we have one implementation right now.

I do not understand the second point of Travis. I think we had this discussion long back on this PR, where I suggested to have `ViewGraph` as input and as output too. But Akshay commented (and I agree with that) that someone has to construct that view-graph data structure too. We cannot do it in two-view plate, so multi-view remains the only choice.

I wrap my head around this by thinking that `ViewGraph` is a graph made of good 2-view relationships (like say rotation-cycle-consistent).

Reply:
> I do not understand the second point of Travis. I think we had this discussion long back on this PR, where I suggested to have ViewGraph as input and as output too. But Akshay commented (and I agree with that) that someone has to construct that view-graph data structure too. We cannot do it in two-view plate, so multi-view remains the only choice.

Creating the ViewGraph is trivial, just fill it with the output of the two-view estimation loop. It creats all of the require data products anyways: `i2Ri1_graph_dict`, `i2Ui1_graph_dict`, `v_corr_idxs_graph_dict`. Then just simply pass these to the ViewGraph.

Reply:
Yeah but like I mentioned that changes the concept of the estimator from something that creates a ViewGraph to something that refines/changes it. This questions the very meaning of a ViewGraph. Is it just a collection of two-view poses or is it two-view noise-free poses that can be passed to averaging? Throughout the design, we have thought of it as the latter.

Reply:
I dont feel too strongly about this, but we would have to change the name of the class as well, and establish a meaning for the ViewGraph. open to comments from @ayushbaid and @johnwlambert